### PR TITLE
W-15885175: split container layers in `mule-tests-runner`

### PR DESCRIPTION
**Pull request by rbourbonnavarro**

### CODE REVIEW

**Review by elrodro83 — CHANGES_REQUESTED** [review]
Separate the 3rd parties dependencies of the mule api in their own layer

File: tests/runner/src/main/java/org/mule/test/runner/api/DependencyResolver.java [hunk 1]
```diff
@@ -199,29 +204,55 @@ public List<Dependency> getDirectDependencies(Artifact artifact, List<RemoteRepo
...
+    return getContainerFiles(node, muleApisLayerDependencies);
+  }
+
+  private List<File> getMuleApisFiles(List<String> excludedFilterPattern)
+      throws DependencyCollectionException, DependencyResolutionException {
+    try {
+      final List<String> apisExcludedFilterPattern = new ArrayList<>(excludedFilterPattern);
+      apisExcludedFilterPattern.add("org.mule.distributions:*:*:*:*");
+      final DependencyFilter dependencyFilter = new PatternExclusionsDependencyFilter(apisExcludedFilterPattern);
+      // TODO - review BOM name and find a way to avoid hardcoding the version
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply tests/runner/src/main/java/org/mule/test/runner/api/DependencyResolver.java:243]: Will tackle this for merging, just a reminder to discuss it.

File: tests/runner/src/main/java/org/mule/test/runner/api/DependencyResolver.java [hunk 2]
```diff
@@ -351,16 +382,19 @@ private List<File> getFiles(DependencyNode node) {
...
    *         are jar urls for third parties.
    */
-  private Pair<List<File>, List<File>> getContainerFiles(DependencyNode node) {
+  private ContainerDependencies getContainerFiles(DependencyNode node, List<File> muleApisLayerDependencies) {
     PreorderNodeListGenerator nlg = new PreorderNodeListGenerator();
     node.accept(nlg);
 
-    LinkedHashSet<File> muleDependencyFiles = new LinkedHashSet<>();
-    LinkedHashSet<File> optDependencyFiles = new LinkedHashSet<>();
+    // TODO - can we rely on this or it might be a subset and that matters?
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply tests/runner/src/main/java/org/mule/test/runner/api/DependencyResolver.java]: Resolved.

File: modules/container/src/main/java11/org/mule/runtime/container/internal/JpmsModuleLayerModuleDiscoverer.java [hunk 3]
```diff
@@ -80,6 +81,17 @@ public List<MuleContainerModule> discover() {
     return discoveredModules;
   }
 
+  private Set<Module> getModules(ModuleLayer layer) {
+    Set<Module> modules = new HashSet<>(layer.modules());
+    for (ModuleLayer parent : layer.parents()) {
+      modules.addAll(getModules(parent));
+    }
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply modules/container/src/main/java11/org/mule/runtime/container/internal/JpmsModuleLayerModuleDiscoverer.java:88]: It's actually fully recursive, it analyzes the parents of the parents all the way up.

**Review by elrodro83 — DISMISSED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**